app/helpers/operatingsystems_helper.rb

```diff
@@ -72,4 +72,12 @@ def os_habtm_family(type, obj)
     result = type.where(:os_family => obj.family)
     result.empty? ? type : result
   end
+
+  def os_default_templates_for_form(os)
+    if os.os_default_templates.any?(&:new_record?)
+      os.os_default_templates.sort_by { |o| o.template_kind.name.downcase }
+    else
+      os.os_default_templates.joins(:template_kind).order('LOWER(template_kinds.name)')
```

Comment:
If there can be a `new_record?`, does that mean that all entries are already in memory anyway? Does even make sense to do it from the database? Worst case this is actually more expensive because you're retrieving all the records again.

Reply:
It's not worse, if there's no new record, the association is reloaded but this time with the right order and without N additional queries loading template_kind and in the right order as a benefit. The worst case is, when there's just one new record and all the rest is already loaded. In that case, rails will load template kind for each os default template as a separate SQL query in sort_by block, but there's not much we can do about it

Reply:
maybe one thing that clarifies it a bit more, not all entries are always in the memory... This is where we build those objects https://github.com/theforeman/foreman/blob/develop/app/controllers/operatingsystems_controller.rb#L25-L33

if the OS has defaults already set, this will behave as standard association without any unpersisted object in the cache